Add character count to edit page

Johennes opened the pull request:
This adds an auto-updating label for the current character count to the bottom of the editing text area.

Closes #6725

This obviously needs more polish. I just mean this to be a basis for discussion for what I requested in #6725 and how to achieve it. Here is an example of how it looks.

![Example](https://imgur.com/8F90zAw)

phlax commented on pootle/templates/editor/units/edit.html:
```diff
@@ -201,6 +201,7 @@
                 alternativeSources: {{ altsrcs|to_js }},
                 fileType: '{{ filetype }}',
                 suggestions: {{ suggestions_dict|to_js }},
+                onChange: function(values) { $('#char-count-{{ unit.id }}').html(values.map(function(e) { return e.length; }).join(' \ ')); },
```
can we move javascript to one of the js files - probs `/static/js/editor/app.js` you can use jquery to listen for change

phlax replied:
also, this prevents the submit button from working

Johennes commented:
Ok, next iteration. This time it should actually work. :)

I couldn't find a good way to handle units with multiple plural forms (= multiple textareas). As a cheap workaround I just joined all the counters with `" / "`. Maybe it's worth rethinking the placement of the counter label in general. I'm currently failing to see a better placement though.

Unrelated question: Would it be preferable to use `"xx characters"` instead of just `"xx"` for the label?

unho commented on pootle/templates/editor/units/edit.html:
```diff
@@ -215,6 +215,13 @@
             {% if unit.istranslated or unit.isfuzzy %}
 	    {% if unit.change.submitted_by and unit.change.submitted_on %}
             <div class="unit-meta">
+              <span id="char-count">
```
Change the id to `js-char-count`.

unho commented on pootle/static/js/editor/app.js:
```diff
@@ -915,6 +915,13 @@ PTL.editor = {
+
+  updateCharacterCount() {
+    let count = ReactEditor.stateValues.map(e => e.length).join(' \ ')
+    $('#char-count').html(count)
```
Travis detects several errors with this Javascript code:

- Must use semicolons at end of relevant lines
- Use `const` instead of `let` because it never gets reassigned

unho commented:
Tested this and it only works for strings that have been previously translated, shouldn't it be better to always show the character count?

I suggest to use `"xx characters"` instead of just `"xx"` for the label because otherwise people won't understand what the count is. Note it needs to be translatable.

Johennes commented:
@unho Thanks for the review. I changed the span's ID and addressed those JS issues. Will look into adding translatable "characters" and not hiding it on untranslated strings now.

Johennes commented:
Ok, I made the label read `"xx Characters"` and prevented it from being hidden for untranslated strings. I also put it on its own line analogous to the editing time and reviewer.

unho commented on pootle/templates/editor/units/edit.html:
```diff
@@ -212,9 +212,15 @@
             <div class="unit-meta">
+              <span id="js-char-count">
+                {% blocktrans with count=unit_values|map_to_lengths|join:" / " %}
+                  {{ count }} Characters
```
Nit: Use `characters` (in lowercase)

unho replied:
Also this needs to be able to handle plurals

Johennes replied:
Yes, so the thing with the plurals is that for messages with more than one text area, the string will read `"12 / 13 / 17 / ... characters"` so I wasn't sure how to properly handle plurals on that one.

unho replied:
IMHO it is ok to always using "characters" if there is more than a textarea, but if there is only a single textarea I would use the plurals.

unho replied:
Or maybe it is better to show the character count for each textarea instead of having the ` / `, but not sure about the impact in the UI. @phlax Any input on this?

unho commented on pootle/static/js/editor/app.js:
```diff
@@ -915,6 +915,15 @@ PTL.editor = {
+
+  updateCharacterCount() {
+    const counters = ReactEditor.stateValues.map(e => e.length).join(' \ ');
+    const format = gettext('%s characters');
```
This needs plurals too:

```javascript
    const counters = ReactEditor.stateValues.map(e => e.length).join(' \ ');
    const string = nt('%s character', '%s characters', counters, [counters]);
    $('#js-char-count').html(string);
```

unho replied:
This might require an import at the top:

```javascript
import { nt } from 'utils/i18n';
```

Johennes commented:
Changed it to correctly pluralize when there is only one text area and always use "characters" when there are multiple.

phlax commented:
lgtm

Johennes commented:
I think this build error has nothing to do with my changes now, or does it? The log mentions Python files I didn't touch.

unho commented:
Yes, the failing tests are unrelated to the changes in this PR.

phlax commented:
@Johennes the last test run has linting errors - that one needs to be fixed - but the other fail is unrelated

Johennes commented:
@phlax I'm sorry but I'm struggling with reading the logs. Could you please point me to where I could see the linting issues?

phlax commented:
@Johennes its fixed - sorry i should have checked

Johennes commented:
Ah, no worries. :)

phlax commented on pootle/core/templatetags/core.py:
```diff
@@ -25,6 +25,13 @@ def to_js(value):
 
+@register.filter
+def map_to_lengths(value):
+    """Maps a list value by replacing each element with its length.
```
this could do with a quick test

Johennes replied:
Added a few basic tests. I hope I put them into the right directory.
